### ISSUE TITLE
Errors while trying to build master on macOS High Sierra 10.13.4

### ISSUE THREAD

**Issue by winzig** [dotnet/corefx#28756]
I'm trying to `./build-native.sh -release` on the master branch of corefx on macOS High Sierra 10.13.4, and it ultimately dies after:

`[ 83%] Built target System.IO.Compression.Native-Static` 

With error `make: *** [all] Error 2`.

[The full output of my build is here.](https://gist.github.com/winzig/ac1c116c761db9dc19e5a15e3a835d56)

Where it seems to go off the rails is when it's compiling pal_networking.c/pal_compiler.h. 

I've followed the [corefx instructions for setting up a development environment for macOS](https://github.com/dotnet/corefx/wiki/Development-environment-setup-for-MacOS):

- openssl 1.0.2o_1 is installed
- cmake version 3.11.0 is installed
- pkg-config version 0.29.2 is installed 
- Xcode 9.3 is installed 

The reason I'm doing `./build-native.sh -release` is because I'm trying to follow [the tip here](https://github.com/dotnet/corefx/issues/28652#issuecomment-377717018) to do a local build of the latest code, to workaround an issue in 2.0.6 on macOS High Sierra 10.13.4. For what it's worth, doing a plain `./build.sh` fails in the same way.

**Comment by kasper3** [comment]
The actual error is:

```
[ 28%] Building C object System.Native/CMakeFiles/System.Native-Static.dir/pal_networking.c.o
/Users/tom/Repositories/corefx/src/Native/Unix/System.Native/pal_networking.c:1042:5: error: declaration shadows a typedef in the global scope [-Werror,-Wshadow]
    c_static_assert(sizeof_member(struct xsocket, so_linger) == 2);
    ^
/Users/tom/Repositories/corefx/src/Native/Unix/Common/pal_compiler.h:19:41: note: expanded from macro 'c_static_assert'
#define c_static_assert(e) typedef char __c_static_assert__[(e)?1:-1]
                                        ^
/Users/tom/Repositories/corefx/src/Native/Unix/System.Native/pal_networking.c:146:1: note: previous declaration is here
c_static_assert(offsetof(struct IOVector, Count) == offsetof(struct iovec, iov_len));
^
/Users/tom/Repositories/corefx/src/Native/Unix/Common/pal_compiler.h:19:41: note: expanded from macro 'c_static_assert'
#define c_static_assert(e) typedef char __c_static_assert__[(e)?1:-1]
                                        ^
1 error generated.
make[2]: *** [System.Native/CMakeFiles/System.Native.dir/pal_networking.c.o] Error 1
[ 29%] Building CXX object System.Security.Cryptography.Native.Apple/CMakeFiles/System.Security.Cryptography.Native.Apple-Static.dir/pal_ecc.cpp.o
make[1]: *** [System.Native/CMakeFiles/System.Native.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 29%] Building C object System.IO.Compression.Native/CMakeFiles/System.IO.Compression.Native.dir/Users/tom/Repositories/corefx/src/Native/AnyOS/brotli/enc/backward_references.c.o
/Users/tom/Repositories/corefx/src/Native/Unix/System.Native/pal_networking.c:1042:5: error: declaration shadows a typedef in the global scope [-Werror,-Wshadow]
    c_static_assert(sizeof_member(struct xsocket, so_linger) == 2);
    ^
/Users/tom/Repositories/corefx/src/Native/Unix/Common/pal_compiler.h:19:41: note: expanded from macro 'c_static_assert'
#define c_static_assert(e) typedef char __c_static_assert__[(e)?1:-1]
                                        ^
/Users/tom/Repositories/corefx/src/Native/Unix/System.Native/pal_networking.c:146:1: note: previous declaration is here
c_static_assert(offsetof(struct IOVector, Count) == offsetof(struct iovec, iov_len));
^
/Users/tom/Repositories/corefx/src/Native/Unix/Common/pal_compiler.h:19:41: note: expanded from macro 'c_static_assert'
#define c_static_assert(e) typedef char __c_static_assert__[(e)?1:-1]
```

Doesn't make much sense as it seems like clang in XCode 9.3 is recognizing a macro "usage" as "declaration"?

**Comment by kasper3** [comment]
@janvorli, should we turn off -Wall in CoreFX, like it was done for CoreCLR? Also clang 4,5,6 support is missing CoreFX (will try to prepare a PR for that).

**Comment by kasper3** [comment]
With clang6 on Ubuntu, there are bunch of errors:

```
[ 35%] Building C object System.Native/CMakeFiles/System.Native-Static.dir/pal_networking.c.o
/corefx/src/Native/Unix/System.Native/pal_networking.c:2048:19: error: implicit conversion changes signedness:
      'enum SocketEvents' to 'int32_t' (aka 'int') [-Werror,-Wsign-conversion]
    sae->Events = GetSocketEvents(events);
                ~ ^~~~~~~~~~~~~~~~~~~~~~~
```

(used `(int)GetSocketEvents(events);` to continue)

```
[ 37%] Building CXX object System.Native/CMakeFiles/System.Native.dir/pal_process.cpp.o
/corefx/src/Native/Unix/System.Native/pal_process.cpp:370:38: error: result of comparison 'uint64_t'
      (aka 'unsigned long') > 18446744073709551615 is always false [-Werror,-Wtautological-type-limit-compare]
    if (value == UINT64_MAX || value > std::numeric_limits<rlim_t>::max())
                               ~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

(used `>=` to continue)

```
[ 68%] Building CXX object System.Net.Http.Native/CMakeFiles/System.Net.Http.Native-Static.dir/pal_easy.cpp.o
/corefx/src/Native/Unix/System.Net.Http.Native/pal_easy.cpp:11:38: error: comparison of two values with different
      enumeration types ('PAL_CURLoption' and 'CURLoption') [-Werror,-Wenum-compare]
static_assert(PAL_CURLOPT_INFILESIZE == CURLOPT_INFILESIZE, "");
              ~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~

(and lots of these)
```

**Comment by janvorli** [comment]
@kasper3 I have actually started to work on fixing these warnings for clang 6 some time ago, but then got distracted by higher priority work for the 2.1 release. I believe I have fixed everything except for the `if (value == UINT64_MAX || value > std::numeric_limits<rlim_t>::max())` - the change to the >= you've made is not correct. The problem is that on some 64 bit platforms, the rlim_t is 64 bit and so the `std::numeric_limits<rlim_t>::max()` is `UINT64_MAX`, thus the warning. But on others (I think it is on OSX), the rlim_t is 32 bit. And in that case, we want the test to pass for all values higher than **but not equal to** `UINT32_MAX` (or maybe it was `INT32_MAX`, I'm not sure).

**Comment by janvorli** [comment]
@winzig as for the OSX build issue - I've updated my xcode to 9.3 and I can repro the issue. I'm looking into it.

**Comment by bartonjs** [comment]
@winzig System.Security.Cryptography.Native.Apple.dylib should have built successfully, though; so you're probably as far along as you need to be. :smile:

**Comment by janvorli** [comment]
Fixed by #28768

**Comment by winzig** [comment]
Thanks so much guys, I can confirm latest from master is building fine here!

**Comment by kasper3** [comment]
@omajid, thanks! also see https://github.com/dotnet/coreclr/issues/16989#issuecomment-379244780

Ubuntu 18.0.4 has llvm-toolchain with the fix in the official repo.

Debian also has the fix in Buster: http://metadata.ftp-master.debian.org/changelogs/main/l/llvm-toolchain-6.0/llvm-toolchain-6.0_6.0-3_changelog (in `-- Sylvestre Ledru <sylvestre@debian.org>  Tue, 24 Apr 2018 23:24:44 +0200`).

Both of these guys don't consider llvm's apt repo as official by maintainers. Something I was informed in both of their devel channels (maybe apt.llvm.org is mostly used by llvm testers).

Which means `apt update && apt install -y liblldb-6.0-dev` in Ubuntu 18.0.4 followed by `find / -name LLDB.h` gives us result now. And a simple `apt install -y lldb` installs lldb v6 (default, without devel package).

RH/Fedora/CentOS and Gentoo etc. already has that cmake patch applied (it was an oversight, which is fixed in 6.0.1 and the exact patch by Pavel is applied by distro packages).
